### DISCUSSION TITLE
Add ICMP NACL rules on a VPC deployed with VPC module from CIS Service Catalog

### DISCUSSION THREAD

**Discussion by klaphi** [gruntwork-io/knowledge-base#160] [Help] [answered]
Hi all,

I am trying to find out howto enable ICMP NACL rules for given network ranges on a VPC created by VPC module of the [CIS catalog](https://github.com/gruntwork-io/terraform-aws-cis-service-catalog).
Possible variables are `nacl_ingress_rules`, `nacl_egress_rules` for the wrapped VPC module or `app_allow_inbound_ports_from_cidr` (CIS NACL module) but the most recent one does not handle egress NACLs. Currently I am not sure what would be the best option and which priority numbers I should use without interfering with the already deployed NACLs.

**Comment by autero1** [comment]
Hi @klaphi ,

What are the source and target subnets for your use case? The default `NACL` rules allow all traffic from within the same "layer" (private -> private) and top-down (public -> private, private -> persistence).

**Comment by klaphi** [comment]
Hi @autero1 ,

my use case is related clients and services hosted in a on-premise data center which is connected to the VPC via TGW and direct connect. We just use private subnets (`create_public_subnets = false`, `create_private_persistence_subnets = false`)

BR
Philipp

**Comment by autero1** [comment, marked as answer]
Thanks for the clarification! You can allow inbound traffic to private subnets with input variable `private_app_allow_inbound_ports_from_cidr` in the `terraform-aws-cis-service-catalog/vpc` module, e.g.:

```
  private_app_allow_inbound_ports_from_cidr = {
    allow_icmp_from_onprem = {
      client_cidr_block = "10.20.30.40/32"
      rule_number = 50
      protocol = "icmp"
      from_port = -1
      to_port = -1
    }
  }
```
Keep the the `rule_number` below 100 to ensure it get higher precedence over the default `vpc` module rules. The CIS Service Catalog rules are all sub 100, but none of the extra CIS rules interfere or overlap with this rule.

WRT to the `egress` configuration, the wrapped `terraform-aws-vpc/vpc-app-network-acls` does not support custom `egress` rules. However, you can use [`terraform-aws-vpc/network-acl-outbound`](https://github.com/gruntwork-io/terraform-aws-vpc/tree/master/modules/network-acl-outbound) to create a custom rule. You can also file a feature request to support extra `egress` rules in `vpc-app-network-acls`.

**Reply by klaphi** [reply]
It seems that the module does not support ICMP. 

````terraform
│ Error: error creating Network ACL (acl-0e2b79d1123405a6c) Egress (false) Rule (50): MissingParameter: The request must contain the parameter icmpTypeCode.type
│ 	status code: 400, request id: eb7ac2a7-b042-4296-b405-7ba44fa8d795
│ 
│   with module.vpc_network_acls.module.vpc_app_network_acls.aws_network_acl_rule.private_app_allow_inbound_from_client_cidr["allow_icmp_from_onprem"],
│   on .terraform/modules/vpc_network_acls.vpc_app_network_acls/modules/vpc-app-network-acls/main.tf line 114, in resource "aws_network_acl_rule" "private_app_allow_inbound_from_client_cidr":
│  114: resource "aws_network_acl_rule" "private_app_allow_inbound_from_client_cidr" {
````

Please check the following [PR](https://github.com/gruntwork-io/terraform-aws-vpc/pull/249)

**Reply by antmordel** [reply]
Hi @klaphi, PR is merged and the code is available in the release [v0.18.10](https://github.com/gruntwork-io/terraform-aws-vpc/releases/tag/v0.18.10).

**Reply by klaphi** [reply]
Hi @antmordel, I have created a [PR](https://github.com/gruntwork-io/terraform-aws-vpc/pull/252) for custom outbound NACLs as well. Please review and merge, because in the current state only half of the story is implemented. Thanks in advance !

**Reply by klaphi** [reply]
Please merge the changes also to CIS sevice catalog, [PR](https://github.com/gruntwork-io/terraform-aws-cis-service-catalog/pull/302) is open.

**Reply by autero1** [reply]
https://github.com/gruntwork-io/terraform-aws-cis-service-catalog/releases/tag/v0.30.2

**Comment by antmordel** [comment]
Hi @klaphi, did you get all sorted out? Could we mark this question as answered? Thank you so much!

**Comment by antmordel** [comment]
Hi @klaphi - just double checking here. Is this issue solved? Many thanks!

**Reply by klaphi** [reply]
Yes this is solved, thanks !
